### PR TITLE
Fix in fisheye calibrate function: issue #5389

### PR DESCRIPTION
**Pull request by elenash**

### CODE REVIEW

File: modules/calib3d/src/fisheye.cpp [hunk 1]
```diff
@@ -794,8 +794,46 @@ double cv::fisheye::calibrate(InputArrayOfArrays objectPoints, InputArrayOfArray
...
     if (K.needed()) cv::Mat(_K).convertTo(K, K.empty() ? CV_64FC1 : K.type());
     if (D.needed()) cv::Mat(finalParam.k).convertTo(D, D.empty() ? CV_64FC1 : D.type());
-    if (rvecs.needed()) cv::Mat(omc).convertTo(rvecs, rvecs.empty() ? CV_64FC3 : rvecs.type());
-    if (tvecs.needed()) cv::Mat(Tc).convertTo(tvecs, tvecs.empty() ? CV_64FC3 : tvecs.type());
+
+    if (rvecs.needed())
+    {
+        if( rvecs.kind() == _InputArray::STD_VECTOR && (rvecs.type() == CV_64FC3 || rvecs.type() == CV_32FC3))
+        {
+            cv::Mat(omc).convertTo(rvecs, rvecs.empty() ? CV_64FC3 : rvecs.type());
```
Comment (paroj) [inline modules/calib3d/src/fisheye.cpp]: I think this also work with Mat and UMat, so you do not need to restrict it to STD_VECTOR

Comment (elenash) [inline modules/calib3d/src/fisheye.cpp]: convertTo fails with segv trying to convert vector of cv:Vec to vector of cv::Mat. So we need to handle this case separately.

Comment (paroj) [inline modules/calib3d/src/fisheye.cpp]: yes, but you only need
```
if(rvecs.kind() == STD_VECTOR_MAT) {
// handle vec mat
} else {
// works for all other cases
}
```

Comment (elenash) [inline modules/calib3d/src/fisheye.cpp]: Ok, I'll fix that, thanks.

**Comment by asmorkalov** [comment]
@vpisarev please look at it.

File: modules/calib3d/src/fisheye.cpp [hunk 2]
```diff
@@ -794,8 +794,46 @@ double cv::fisheye::calibrate(InputArrayOfArrays objectPoints, InputArrayOfArray
...
+        {
+            rvecs.create((int)objectPoints.total(), 1, CV_64FC3);
+            for( int i = 0; i < (int)objectPoints.total(); i++ )
+            {
+                rvecs.create(3, 1, CV_64F, i, true);
+                Mat rv = rvecs.getMat(i);
+                *rv.ptr<Vec3d>(0) = omc[i];
+            }
+        }
+        else if(rvecs.empty() || rvecs.type() == CV_64FC3 || rvecs.type() == CV_32FC3)
```
Comment (vpisarev) [inline modules/calib3d/src/fisheye.cpp]: it's not very good style. The function should not rely on the actual type of pre-allocated rvecs.
The correct code would be:
<pre>
if  (rvecs.kind() == _InputArray::STD_VECTOR_MAT)  { ... }
else
{
     cv::Mat(omc).convertTo(rvecs, rvecs.fixedType() ? rvecs.type() : CV_64FC3);
}

File: modules/calib3d/src/fisheye.cpp [hunk 3]
```diff
@@ -794,8 +794,46 @@ double cv::fisheye::calibrate(InputArrayOfArrays objectPoints, InputArrayOfArray
...
+        {
+            tvecs.create((int)objectPoints.total(), 1, CV_64FC3);
+            for( int i = 0; i < (int)objectPoints.total(); i++ )
+            {
+                tvecs.create(3, 1, CV_64F, i, true);
+                Mat tv = tvecs.getMat(i);
+                *tv.ptr<Vec3d>(0) = Tc[i];
+            }
+        }
+        else if( tvecs.empty() || tvecs.type() == CV_64FC3 || tvecs.type() == CV_32FC3)
```
Comment (vpisarev) [inline modules/calib3d/src/fisheye.cpp]: same problem

File: modules/calib3d/src/fisheye.cpp (hunk 2 above)
Comment (elenash) [inline modules/calib3d/src/fisheye.cpp]: Fixed, thanks

File: modules/calib3d/src/fisheye.cpp (hunk 3 above)
Comment (elenash) [inline modules/calib3d/src/fisheye.cpp]: Fixed

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**